framework/src/tinyalsa/tinyalsa.c

```diff
@@ -733,6 +733,7 @@ int pcm_close(struct pcm *pcm)
 	}
 
 	if (pcm == &bad_pcm) {
+		free(pcm);
```

Comment:
I think we don't have to keep bad_pcm, original tinyalsa asign pcm as bad_pcm if mem allocation failed.
it returns bad_pcm if malloc failed in pcm_open. but we return null instead of bad pcm.
Could you please remove bad_pcm structure & remove this condition too?

Reply:
Could you also consider the case when open call fails in pcm_open, then allocated pcm struct is returned which has fd=-1 which makes it equivalent ot bad_pcm? Should I return NULL in that case too?

Reply:
In my opinion, free at pcm_open and return null, or this seems to work.

Reply:
If it allocated but fd is invalid, in this case it will be free at pcm_close.
In player level, it should check pcm_is_ready, and if its return value is false, it should call pcm_close immediately.

Reply:
So, is this change fine?

Reply:
I mean remove bad_pcm structure & related code will be fine.

Reply:
If I remove bad_pcm, then the UNREGISTERMQ, FREEBUFFER and RELEASE ioctl operations will be called which may lead to errors because neither buffer or message queue were allocated. 
Or do you want me to not use bad_pcm and directly check the value of fd?

Reply:
Yes. then 'if (pcm->fd >= 0)' can be removed.
Actually bad_pcm is static structurem and we do not set address of that structure to pcm anywhere.

Reply:
Ok, I understood